sourcecred/sourcecred
# Explorer Home: Fix display of Average Grain in Activity table

# Description
In fixing other issues, I noticed that the calculation of Average Grain in the Activity Table on Explorer Home was incorrect (always 0). The TLDR is that Grain is saved as a very large, high-precision string and when dividing the number of grain / the number of users, the number of users was being converted into a very large integer as well, which made the precision drop by 18 decimal places and therefore appear as 0. (E.g. what should have been, say 450 was .000000000000000450, which displayed as 0).

The fix is to simply allow the `div()` function to accept a number as the divisor since that's converted directly into a BigInt. In fact, I'd argue that the divisor should **only** be a number, and not a Grain, because the conceptual idea is that you're dividing a quantity of grain into a certain number of parts, and if the number of parts is also padded by 18 decimal places, you'll lose those places in the division such that the quotient becomes stripped of its 18 decimal places of precision (as happened above).

One drawback of this fix is that not **any** number is acceptable, as BigInt only accepts integers. Any non-integer number passed in would throw an error. I looked into Flow typing to handle this, but it does not seem to be able to specify "integers only".

An alternative fix would be to convert the number of users to a string, then create a grain value from that string using `fromString`, which does not add the 18-places of precision like `fromInteger` does. In fact, the test for the `div()` function does just that. However, `fromString(String(someNumber))` is messy and very unintuitive when a `fromInteger()` function exists, even though that's not the right choice in this case either.

Also, all of the above arguments apply equally to the `mul()` function as well, with the exception that multiplication is associative so it technically shouldn't matter which factor is the Grain and which is the integer, but it would make sense to keep the same signature as `div()` for consistency.

Feedback and improved solutions are welcome.

# Test Plan
* Load page
* Verify Average Grain is indeed the average for the participants shown
* Filter list down to only a few entries
* Verify Average Grain is updated to reflect the average for only the participants shown



┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200032326400385/1200046196345722) by [Unito](https://www.unito.io/learn-more)
